gobblin-core-base/src/main/java/gobblin/crypto/RotatingAESEncryptor.java

```diff
@@ -0,0 +1,168 @@
+package gobblin.crypto;
+
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.codec.binary.Base64OutputStream;
+
+import gobblin.writer.StreamEncoder;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import javax.xml.bind.DatatypeConverter;
+
+
+/**
+ * Implementation of an encryption algorithm that works in the following way:
+ *
+ * 1. A credentialStore is provisioned with a set of AES keys
+ * 2. When wrapOutputStream() is called, an AES key will be picked at random
+ * 3. A header will be written [keyId][ivLength][base64 encoded iv]
+ * 4. Ciphertext will be base64 encoded and written out
+ */
+public class RotatingAESEncryptor implements StreamEncoder {
+  private static final int AES_KEY_LEN = 16;
+
+  private final SecureRandom random;
+  private final CredentialStore credentialStore;
+  private boolean headerWritten;
+  private OutputStream origStream;
+  private Cipher cipher;
+  private String base64Iv;
+  private KeyRecord currentKey;
+  private List<KeyRecord> keyRecords;
+
+  /**
+   * Create a new encryptor
+   * @param credentialStore Credential store where keys can be found
+   */
+  public RotatingAESEncryptor(CredentialStore credentialStore) {
+    this.credentialStore = credentialStore;
+    this.random = new SecureRandom();
+    this.headerWritten = false;
+  }
+
+  @Override
+  public OutputStream wrapOutputStream(OutputStream origStream) {
+    this.origStream = origStream;
+
+    rekey();
+    final Base64OutputStream base64OutputStream = new Base64OutputStream(origStream);
+    final CipherOutputStream encryptedStream = new CipherOutputStream(base64OutputStream, cipher);
```

Comment:
I originally wrote it that way - I'm not sure if I'm being overly paranoid, but I figured it is possible that  there will be cases later when the full set of OutputStreams have not been chained together when wrapOutputStream() is called. 

Not writing the header until the first write() call is done means we aren't actually doing I/O until a specific write is requested which I think is the safer way to go. Otherwise it's not super intuitive that calling wrapOutputStream(os) may actually end up doing I/O behind the scenes